Assembler.Base/MiddleFrameHandler.cs

```diff
@@ -0,0 +1,67 @@
+﻿using System;
+using Assembler.Core;
+using Assembler.Core.Entities;
+
+namespace Assembler.Base
+{
+    public class MiddleFrameHandler<TFrame, TMessage> : BaseFrameHandler<TFrame, TMessage>
+        where TFrame : BaseFrame
+        where TMessage : BaseMessageInAssembly
+    {
+        private readonly ILogger _logger;
+
+        public MiddleFrameHandler(ITimeBasedCache<TMessage> cache,
+            IFactory<TFrame, string> identifierFactory, ICreator<TMessage> assembledMessageCreator,
+            IMessageEnricher<TFrame, TMessage> enricher, ILoggerFactory loggerFactory) : base(cache, identifierFactory,
+            enricher, assembledMessageCreator)
+        {
+            _logger = loggerFactory.GetLogger(this);
+        }
+
+        public override void Handle(TFrame frame)
+        {
+            string identifier;
+
+            try
+            {
+                identifier = GetIdentifier(frame);
+            }
+            catch (Exception e)
+            {
+                _logger.Error($"There was an error while getting an identifier out of the frame [{frame.Guid}], " +
+                              $"it won't be used it the assembling process \n {e}");
+                return;
+            }
+
+            TMessage message = GetOrCreateMessageInAssembly(identifier);
+
+            MessageEnricher.Enrich(frame, message);
+
+            _logger.Debug(
+                $"Enriched [{message.Guid}] with the frame [{frame.Guid}] ");
+
+            Cache.Put(identifier, message);
+        }
+
+        private TMessage GetOrCreateMessageInAssembly(string identifier)
+        {
+            if (Cache.Exists(identifier))
+            {
+                return Cache.Get<TMessage>(identifier);
```

Comment:
You should set the releaseReason here also, shouldnt you? If so, set it in the Handle method above, after calling `GetOrCreateMessageInAssembly` (and then you can remove line 59: `message.MiddleReceived = true;`)

Reply:
That's what the comment is about

Reply:
Sorry, I was wrong. I meant you need to set MiddleRecieved here too, right?

Reply:
f

Reply:
Cool, but you added this:
`// TODO: Test that it happens both from in cache and in a new one`

did you add it?

Reply:
f